core/lib/object_store/src/gcs_public.rs

```diff
@@ -0,0 +1,163 @@
+use std::time::Duration;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use http::StatusCode;
+use reqwest::Client;
+
+use crate::{raw::BoxedError, Bucket, ObjectStore, ObjectStoreError};
+
+#[derive(Debug)]
+pub struct PublicReadOnlyGoogleCloudStorage {
+    gcs_base_url: String,
+    bucket_prefix: String,
+    max_retries: u16,
+    client: Client,
+    backoff_seconds: u64,
+}
+
+impl PublicReadOnlyGoogleCloudStorage {
+    pub fn new(
+        gcs_base_url: String,
+        bucket_prefix: String,
+        max_retries: u16,
+        timeout: u64,
+        backoff_seconds: u64,
+    ) -> Self {
+        let client = Client::builder()
+            .timeout(Duration::from_secs(timeout))
+            .build()
+            .expect("Unable to build PublicReadOnlyGoogleCloudStorage http client from config");
+        Self {
+            gcs_base_url,
+            bucket_prefix,
+            max_retries,
+            client,
+            backoff_seconds,
+        }
+    }
+}
+
+#[async_trait]
+impl ObjectStore for PublicReadOnlyGoogleCloudStorage {
+    async fn get_raw(&self, bucket: Bucket, key: &str) -> Result<Vec<u8>, ObjectStoreError> {
+        let mut last_error: Option<anyhow::Error> = None;
+        let retry_count = self.max_retries;
+        for retry_number in 0..retry_count {
```

Comment:
Does `retry_count = 0` means an instant panic? I'd suggest to check `max_retries` is not equal to zero in the constructor then. Alternatively, `max_retries = 0` may be defined to mean "a single attempt"; note that an adjacent `GoogleCloudStorage` implementation uses this definition, so it would be more consistent to use it here as well.